### ISSUE TITLE
Suggestion: Looking at pyTectonics

### ISSUE THREAD

**Issue by esampson** [Mindwerks/worldengine#123]
I just stumbled over this project yesterday: http://pytectonics.sourceforge.net/. While it doesn't run completely correctly on my system (I can't interact with the windows) I thought that it might bear looking at. If we could remove its reliance on VPython it has an underlying plate tectonic system that seems pretty good that operates on a spherical planet that could replace platec. It also looks like it has an underlying coordinate system that could be used for things like river routing, erosion, and rainfall that would allow that activity to work correctly up at the poles.

It would probably result in .world files that aren't compatible with the older .world files since they would use a different coordinate system but it seems like translating those worlds into the PNG images that are currently being exported would be fairly trivial.

**Comment by ftomassetti** [comment]
I saw this project some time ago. Personally I think that platec provides better results (but not on a sphere, that is true). In my opinion to move to a spherical simulation would require a massive amount of work and I do not think we can manage in a reasonable time without new contributors helping us.

**Comment by tcld** [comment]
I noticed that it says
"PyTectonics is no longer maintained. There is a replacement, which you can see here: davidson16807.github.io/tectonics.js."
on the project page at
http://sourceforge.net/projects/pytectonics/

The new project is at
https://github.com/davidson16807/tectonics.js

A nice overview over possible parameters can be found here:
https://github.com/davidson16807/tectonics.js/blob/master/scripts/model/World.js

Even without the constant references to scientific papers this would look very, very impressive to me. How much work would you estimate is necessary to use the tectonics-project as an alternative to platec (I assume that flat maps are still interesting in many cases)?

**Comment by ftomassetti** [comment]
well, it is written in Javascript so we should use some Javascript engine with Python bindings to run it.
About the effort we could either:
1) Project the world on the rectangular map after the tectonics simulation: in this case all the other step would remain the same and this would be one possible alternative to platec
2) Run all the simulation on a sphere: that basically means rewriting/adapat all the simulations (temperature simulation, rain simulation, erosion, etc.)

If we start with 1 we can move to 2 later in the future I guess.

**Comment by tcld** [comment]
Yes, I thought about that. How "good" would the simulations be if they were acting on a projected sphere? Would there be some kind of "information pile-up" at the poles?

**Comment by ftomassetti** [comment]
I am not sure, there are a few generators how there which use projections and you can look at the Earth on a map: there are some distorsions possible but we can reduce the impact of some of them. For example after tectonics simulation we shift the world to reduce the amount of land at the borders (so it is very rare to have continents which are wrapped E-W or N-S). We could think of something similar to reduce these distorsions.

**Comment by psi29a** [comment]
Do any of these libraries offer the same level of features (simulations) as our current Platec library?

When it comes to speed, C/C++ will always beat out Python, Javascript and etc.

**Comment by tcld** [comment]
I might take a closer look at this later and then report back.

**Comment by tcld** [comment]
I read through some of the code-files and so far things look fairly simple (not necessarily in the way where it would be simple to make use of that code from within Worldengine). What are some of the nicer features that platec offers? I suspect that it might be a little more evolved than this Javascript-code, but I have only looked at very little of the platec code.

**Comment by tcld** [comment]
Has anybody ever looked into making platec create spherical worlds? If one could represent a spherical world as something like a linked list (and I think that should be possible), then were to feed that list into platec instead of its normally used matrices, then replaced all kinds of "standard" loops with iterator-loops and *then* moved all possible border-checks from generating code into the data-structure itself...

...the same platec-code could be fed different maps to populate. The biggest difference between a flat and a spherical one is probably the "looping element" that a sphere comes with. It has to be handled once, and then every border-check will be obsolete forever.

That is the roughest of rough ideas, and that is probably why it sounds so very possible to me. @ftomassetti , you do know the platec-code in detail, right? Can you imagine it using an iterable linked list to carry data instead of an array/a matrix? There is a lot of code I haven't seen yet.

**Comment by ftomassetti** [comment]
Unfortunately I do not know platec in details: it was built as a project for the master thesis of a Finnish student (available here: http://www.theseus.fi/handle/10024/40422). I just took it, added tests, split some methods and classes and tried to refactor some parts but it is not a trivial piece of code and there are not so many comments so I cannot really say I understand it very well. That and the fact that is difficult to test make very difficult to change it. I did some changes in the past, for example I made possible to generate rectangular maps instead of just square maps but it was a major effort.

I cannot say that I am an expert in platec but I think that a linked list would be very slow compared to an array because we want to access points at an arbitrary index all the time. Many operations are not loop but they require to verify if there is a plate in a given point and modify the crust in that given point and in the points around. With an array those are constant-time operations while with a linked list they becomes linear time operation I guess.

I think it can be doable to make platec works on a sphere but it would require a lot of work and determination. Perhaps we can use the fact that C++ let you override operators to change subscript operators: my goal to reduce the effort of the change would be to keep using rectangular coordinates to access what would appear to be bi-dimensional arrays and do all the magic inside the implementation of these objects. I am not sure this is feasible but if we can do that we would not have to rewrite all the code.

**Comment by tcld** [comment]
Working on a flat map we currently have to be careful at the borders in whatever method is run. Should we take that flat map as the base for a sphere, the borders, I assume, will become much, much difficult to handle.

Yes, a linked list is probably slow. I don't know a good type of data to store a polygonized sphere in, though.

I will keep thinking about it. Currently I am trying to wrap my head around numpy, after that I would like to get ocean_level working. So there is enough to do already.^^

**Comment by psi29a** [comment]
I honestly like c++ (for many reasons), one being that there is almost always a library out there that does what you want.

We are still contact the original Platec developer and he is working on adding more features in addition to merging what @ftomassetti has developed so that they are not two desperate projects.

I think it would be a good idea to collaborate with him. I'm sure he is willing to answer questions about the code and take feature requests.

**Comment by tcld** [comment]
It's good to know that he is still active. I will take a look into that route, should I find the time to properly approach spherical worlds. And maybe there are even more advanced plate-simulation libraries out there, who knows...

**Comment by psi29a** [comment]
If you want, I can use your email address you use in your commits to send you his email.

**Comment by tcld** [comment]
Thank you. I will come back to it, if necessary. I want to focus on the simpler things first.

**Comment by psi29a** [comment]
Going to close this thread; I believe any additional work should be done as part of a feature request in PyPlatec (sphere).
